### PR TITLE
[sweet][kotlin] Refactor conversion system

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

I wasn't very happy with the current conversion system. So I rewrote it. 

# How

Previously, in each method call, we checked how we should convert the argument. It was working. However, this solution wasn't particularly fast due to huge runtime overhead. I thought we could do it better by storing a converter for each argument type. My solution shouldn't increase memory usage a lot, cause converters for a common type are reusable. Moreover, I only store converters for methods that were invoked at least once. 

Also, I saw that `Dynamic` should be recycled after each usage. In some cases, the object comes from the pool. 

And finally, I fix a bug with the typed array conversion. 

# Test Plan

- cellular ✅
- unit tests ✅

### CODE REVIEW

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/records/RecordTypeConverter.kt [hunk 1]
```diff
@@ -3,26 +3,28 @@ package expo.modules.kotlin.records
...
+// TODO(@lukmccall): create all converters during initialization
+class RecordTypeConverter<T : Record>(
+  private val converterProvider: TypeConverterProvider,
+  val type: KType,
+) : TypeConverter<T>(type.isMarkedNullable) {
   private val objectConstructorFactory = ObjectConstructorFactory()
 
-  override fun canHandleConversion(toType: KClassTypeWrapper): Boolean =
-    Record::class.isSuperclassOf(toType.classifier)
+  override fun notOptionalConvert(value: Dynamic): T {
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/records/RecordTypeConverter.kt:21]: Nitpick 😅
1. "not-optional" sounds less natural than "non-optional" when it's an adjective. Not only here, but everywhere in this PR code
2. IMO inverted word order is more descriptive (the converted value is non-optional, not the conversion itself)

```suggestion
  override fun convertNonOptional(value: Dynamic): T {
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  return converter.convert(value)
+}
+
+object TypeConverterProviderImpl : TypeConverterProvider {
+  private val cachedConverters = createCashedConverters(false) + createCashedConverters(true)
+
+  override fun obtainTypeConverter(type: KType): TypeConverter<*> {
+    cachedConverters[type]?.let {
+      return it
+    }
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:42]: I read the PR's description, but I still don't get it. Why do you cache only some types here, while others are returned later in this function each time it's called?

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/AnyType.kt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package expo.modules.kotlin.types
+
+import com.facebook.react.bridge.Dynamic
+import kotlin.reflect.KType
+
+fun KType.toAnyType(): AnyType {
+  return AnyType(this)
+}
```
Comment (barthap) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/AnyType.kt:8]:
```suggestion
fun KType.toAnyType() = AnyType(this)
```
Up to you

**Review by barthap — APPROVED** [review]
LGTM, just one question 😉

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt (hunk 2 above)
Comment (lukmccall) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:42]: Ok, it isn't the code that I referred to in the description. The type converter is cached in the `AnyType` class. 

However, for some basic types like we don't have to create a multiple converter object for each prop, cause the conversion looks always the same. But, if we are talking about conversion to list for instance. We can't just create one converter. Conversion to the list of int will be different than the list of strings. 

So to sum everything up, for basic types, we share the same instance of converter across the whole application. For more complicated types, we create a dedicated converter for each prop. After that, we store the reference to the converter in AnyType. So we don't have to obtain a converter before each call.

Comment (barthap) [reply packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/types/TypeConverterProvider.kt:42]: Yeah, that makes sense, thanks for the explanation

**Pull request merged**